lambda_src/lambda_function.py

```diff
@@ -97,15 +102,115 @@ def sync_flow(event: Any, context: Any = None) -> Dict[Text, Any]:
     LOG.debug('Destination header not found in a POST and hence using sync_flow().')
     headers = event['headers']
     req_body = loads(event['body'])
+    start_time = timer()
 
+    destination_driver = None
     batch_id = headers[BATCH_ID_HEADER]
     write_uri = headers.get('write-uri')
+
     LOG.debug(f'sync_flow() received destination: {write_uri}.')
 
     if write_uri:
         destination_driver = import_module(
             f'geff.drivers.destination_{urlparse(write_uri).scheme}'
         )
+
+    while True:
+        data = get_data_from_lock(batch_id)
+        if data is None:  # request hasn't been initialized
+            break
+        elif data != LOCKED:  # if false, the response is yet to be written
+            return data
+
+    open_lock(batch_id)  # initilaze the request
+
+    res_data = process_request(
+        req_body, headers, event, batch_id, write_uri, destination_driver
+    )
+
+    # Write data to s3 or return data synchronously
+    if write_uri:
+        response = destination_driver.finalize(  # type: ignore
+            write_uri, batch_id, res_data
+        )
+    else:
+        data_dumps = dumps({'data': res_data}, default=str)
+        response = {
+            'statusCode': 200,
+            'body': b64encode(compress(data_dumps.encode())).decode(),
+            'isBase64Encoded': True,
+            'headers': {'Content-Encoding': 'gzip'},
+        }
+        end_time = timer()
+        if response and (end_time - start_time) > 20:
+            LOG.debug('Storing the response in DynamoDB.')
+            try:
+                close_lock(batch_id, response)  # write the response
+            except ClientError as ce:
+                if ce.response['Error']['Code'] == 'ValidationException':
```

Comment:
It's the exception returned when the size limit of an itme is hit. The error message:
> An error occurred (ValidationException) when calling the PutItem operation: Item size has exceeded the maximum allowed size

The rest of the flow continues, as it would earlier. A response is only returned if it's recieved within 30 seconds.

Reply:
Will this leave the lock in a permanently open state?

Reply:
Not permanently actually, the TTL of 30m will cause the item to be deleted afterwards, DynamoDB deletes items within 48h of expiration.

Reply:
per conversation, this is too long -- if we get a 1mb response in 40s, we don't want the user to have to wait 10m to get a "your request has timed out" error